### PR TITLE
[WFLY-13700] Add ability to configure stale transaction timeout for l…

### PR DESCRIPTION
**Pull request by tadamski**
https://issues.redhat.com/browse/WFLY-13700

### CODE REVIEW

File: transactions/src/main/resources/org/jboss/as/txn/subsystem/LocalDescriptions.properties [hunk 1]
```diff
@@ -87,3 +87,5 @@ transactions.commit-markable-resource.jndi-name=JNDi name of this CMR resource
 transactions.commit-markable-resource.xid-location=Name of database table associated to this CMR resource
 transactions.commit-markable-resource.batch-size=Batch size for this CMR resource
 transactions.commit-markable-resource.immediate-cleanup=Immediate cleanup associated to this CMR resource
+
+transactions.stale-transaction-time=The time for which transaction is held in memory after it is being completed.
```
Comment (ochaloup) [inline transactions/src/main/resources/org/jboss/as/txn/subsystem/LocalDescriptions.properties:91]: only for transactions which contains the remote enlistments (aka. which contains the EJB remote call). This should be emphasized in the description as for any other transaction (aka. the txn with DB) the transaction is removed imediatelly after it's committed.

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+            <xs:element name="recovery-environment" type="recovery-environment" maxOccurs="1"/>
+            <xs:element name="coordinator-environment" type="coordinator-environment" minOccurs="0" maxOccurs="1"/>
+            <xs:element name="object-store" type="object-store" minOccurs="0" maxOccurs="1"/>
+            <xs:element name="jts" type="jts-Type" minOccurs="0" maxOccurs="1"/>
+            <xs:choice minOccurs="0" maxOccurs="1">
+                <xs:element name="use-journal-store" type="use-journal-store-Type" minOccurs="0" maxOccurs="1"/>
+                <xs:element name="jdbc-store" type="jdbc-store-Type" minOccurs="0" maxOccurs="1"/>
+            </xs:choice>
+            <xs:element name="commit-markable-resources" type="cmr-resources-Type" minOccurs="0" maxOccurs="1"></xs:element>
+            <xs:element name="stale-transaction-time" type="stale-transaction-time" minOccurs="0" maxOccurs="1"/>
```
Comment (ochaloup) [inline transactions/src/main/resources/schema/wildfly-txn_6_0.xsd:55]: have you been considering to create rather some "WFTC element" which could gathering all the WFTC setup?
Just thinking - until now the setup here was mostly about Narayana. This one is specific for WFTC - so just thinking.

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+    </xs:complexType>
+
+    <xs:complexType name="stale-transaction-time">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The time after which completed transactions are removed from the memory.
```
Comment (ochaloup) [inline transactions/src/main/resources/schema/wildfly-txn_6_0.xsd:345]: the  same comment as for description in cli - this is only for transactions with remote enlistments (remote EJB XAResource enlisted)

**Review by ochaloup — CHANGES_REQUESTED** [review]
I would like ask for adjusting the description of the stale property and then I have one question about where the settings belong to in the subsystem.

File: transactions/src/main/resources/org/jboss/as/txn/subsystem/LocalDescriptions.properties (hunk 1 above)
Comment (tadamski) [reply transactions/src/main/resources/org/jboss/as/txn/subsystem/LocalDescriptions.properties]: fixed

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd (hunk 2 above)
Comment (tadamski) [reply transactions/src/main/resources/schema/wildfly-txn_6_0.xsd]: fixed

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd (hunk 3 above)
Comment (tadamski) [reply transactions/src/main/resources/schema/wildfly-txn_6_0.xsd]: fixed

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+    </xs:complexType>
+
+    <xs:complexType name="client">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The time after which completed transactions that contain remote enlistments are removed from the memory.
```
Comment (ochaloup) [inline transactions/src/main/resources/schema/wildfly-txn_6_0.xsd:345]: this annotation element should be part of the `<xs:attribte ...` element, or shouldn't?

**Review by ochaloup — CHANGES_REQUESTED** [review]
Thanks for the changes. They addressed my points perfectly. I'm just adding one more as I think the documentation of the property should be inside of the attribute now.

File: transactions/src/main/resources/schema/wildfly-txn_6_0.xsd (hunk 4 above)
Comment (tadamski) [reply transactions/src/main/resources/schema/wildfly-txn_6_0.xsd]: fixed

**Review by ochaloup — APPROVED** [review]
Seems fine to me. Thanks

**Comment by bstansberry** [comment]
Transformation tests in TransactionSubsystemTestCase would be good.

**Comment by tadamski** [comment]
@bstansberry I wasn't able to make transformation test work properly IMO because of the bug so I have created https://github.com/wildfly/wildfly-core/pull/4302.

**Comment by bstansberry** [comment]
@tadamski https://issues.redhat.com/browse/WFCORE-5089 is fixed in the version of core WF master is using now. Thanks for that.

**Comment by tadamski** [comment]
rebased

**Review by darranl — CHANGES_REQUESTED** [review]
This PR is for a Feature Request but I don't see a corresponding RFE for the changes?

Also when bumping a subsystem version can you please separate the bump into a separate commit to the changes, it is not possible to use the GitHub UI to see things like the changes to the schema.

**Comment by bstansberry** [comment]
@tadamski sorry -- i forgot that we were going to handle this without all the kanban process. 

Can you rebase this and ping me so we can get it in? Thanks.

File: transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionSubsystemRootResourceDefinition.java [hunk 5]
```diff
@@ -281,6 +281,13 @@
             .setAllowExpression(true)
             .setRequires(CommonAttributes.USE_JDBC_STORE).build();
 
+    public static final SimpleAttributeDefinition STALE_TRANSACTION_TIME = new SimpleAttributeDefinitionBuilder(CommonAttributes.STALE_TRANSACTION_TIME, ModelType.INT, true)
+            .setValidator(new IntRangeValidator(0))
+            .setMeasurementUnit(MeasurementUnit.SECONDS)
+            .setDefaultValue(new ModelNode().set(600))
+            .setFlags(AttributeAccess.Flag.RESTART_NONE)
```
Comment (bstansberry) [inline transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionSubsystemRootResourceDefinition.java:288]: This should be RESTART_ALL_SERVICES -- ReloadRequiredWriteAttributeHandler handles the writes.

File: transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionTransformers.java [hunk 6]
```diff
@@ -39,6 +40,8 @@
     static final ModelVersion MODEL_VERSION_EAP70 = ModelVersion.create(3, 0);
     static final ModelVersion MODEL_VERSION_EAP71 = ModelVersion.create(4, 0);
     static final ModelVersion MODEL_VERSION_EAP72 = ModelVersion.create(5, 0);
+    static final ModelVersion MODEL_VERSION_5_1_0 = ModelVersion.create(5, 1);
+    static final ModelVersion MODEL_VERSION_EAP73 = ModelVersion.create(6, 0);
```
Comment (bstansberry) [inline transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionTransformers.java:44]: This would be MODEL_VERSION_EAP74

**Review by bstansberry — CHANGES_REQUESTED** [review]
I made a couple comments re things I just noticed.

Also, please add coverage of this to TransactionSubsystemTestCase. Perhaps that would be adding the relevant config to full-expressions.xml and adding a full-6.0.0.xml file and a transformer test method that uses it.

File: transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionSubsystemRootResourceDefinition.java (hunk 5 above)
Comment (tadamski) [reply transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionSubsystemRootResourceDefinition.java]: fixed

File: transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionTransformers.java (hunk 6 above)
Comment (tadamski) [reply transactions/src/main/java/org/jboss/as/txn/subsystem/TransactionTransformers.java]: fixed

**Comment by tadamski** [comment]
hi @bstansberry, this PR should work fine now

**Review by bstansberry — APPROVED** [review]

**Comment by mchoma** [comment]
> This PR is for a Feature Request but I don't see a corresponding RFE for the changes?

I agree here.

**Comment by mchoma** [comment]
> @tadamski sorry -- i forgot that we were going to handle this without all the kanban process.

Why it was decided so? I see model changes in MR so I would expect that we need documentation and tests for it.

**Pull request merged**